### PR TITLE
Exercise 12

### PR DESCRIPTION
**Pull request by cc1000**

### CODE REVIEW

File: build_scripts/deploy.sh [hunk 1]
```diff
@@ -1 +1,2 @@
+echo "Calling webhook to trigger app restart: $1"
```
Comment (BenRamchandani) [inline build_scripts/deploy.sh]: The webhook is secret, it contains the deployment key for the App Service! Travis logs are public e.g. https://travis-ci.com/github/cc1000/DevOps-Course-Starter/builds/227703742#L2273

Remember to rotate the password once this is fixed (on the Azure portal go to the App Service -> Container Settings -> FTPS credentials -> Application scope).

File: outputs.tf [hunk 2]
```diff
@@ -0,0 +1,7 @@
+output "webapp_url" {
+  value = "https://${azurerm_app_service.main.default_site_hostname}"
+}
+
+output "cd_webhook" {
+  value = "https://${azurerm_app_service.main.site_credential[0].username}:${azurerm_app_service.main.site_credential[0].password}@${azurerm_app_service.main.name}.scm.azurewebsites.net/docker/hook"
```
Comment (BenRamchandani) [inline outputs.tf:6]: Should be marked as `sensitive = true` so it doesn't show up in CI/CD logs

**Review by BenRamchandani — CHANGES_REQUESTED** [review]
The Terraform itself looks good, but your CI script exposes deployment credentials at the moment, so we need to fix that.

Can you also enable authentication on the Terraformed app? You might need to make a new GitHub OAuth app. Don't forget to set the Flask SECRET_KEY.

One other unrelated thing: your .env.template file specifies `FLASK_APP=app`, I think it should be `FLASK_APP=src/app`.

File: build_scripts/deploy.sh (hunk 1 above)
Comment (cc1000) [reply build_scripts/deploy.sh]: Done. Auth also now enabled for Azure-hosted app.

File: outputs.tf (hunk 2 above)
Comment (cc1000) [reply outputs.tf:6]: Done.

**Comment by cc1000** [comment]
Followed up on comments above. Going ahead with merge but please let me know if any other changes are needed :)

File: variables.tf [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  description = "The prefix used for all resources in this environment"
+}
+
+variable "location" {
+  description = "The Azure location where all resources in this deployment should be created"
+  default     = "uksouth"
+}
+
+variable "app-authentication-disabled" { default = true }
+variable "app-flask-secret-key" {}
```
Comment (hugh-emerson) [inline variables.tf:11]: These aren't being leaked currently, but variables can also have `sensitive = true` set to ensure they don't get logged.

**Review by hugh-emerson — COMMENTED** [review]
Looks good.

I've left one minor comment, and also your OAuth is working but could be improved:
- You are redirecting to the http version of the site. Should be https
- Your `get_roles` function means that unknown users (like me) can't even view the app :( The intention of the module 10 exercise was that unknown users are still readers.

But anyway, merge when you're ready

**Pull request merged**